### PR TITLE
SQL: Implement GREATEST and LEAST functions

### PR DESCRIPTION
**Pull request by matriv**
Add GREATEST(expr1, expr2, ... exprN) and LEAST(expr1, expr2, exprN)
functions which are in the family of CONDITIONAL functions.

Implementation follows PostgreSQL behaviour, so the functions return
`NULL` when all of their arguments evaluate to `NULL`.

Renamed `CoalescePipe` and `CoalesceProcessor` to `ConditionalPipe` and
`ConditionalProcessor` respectively, to be able to reuse them for
`Greatest` and `Least` evaluations. To achieve that `ConditionalOperation`
has been added to differentiate between the functionalities at execution
time.

Closes: #35878

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java [hunk 1]
```diff
@@ -140,9 +139,6 @@ protected PhysicalPlan rule(ProjectExec project) {
                         if (pj instanceof ScalarFunction) {
                             ScalarFunction f = (ScalarFunction) pj;
                             processors.put(f.toAttribute(), Expressions.pipe(f));
-                        } else if (pj instanceof In) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: Leftover: `In` is already extending `ScalarFunction`.

**Review by costin — CHANGES_REQUESTED** [review]
Looks good however the optimizer rule could be improved as well as the class hierarchy.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/Conditionals.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    static Object leastOnInput(List<Processor> processors, Object input) {
+        List<Object> values = new ArrayList<>(processors.size());
+        for (Processor processor : processors) {
+            values.add(processor.process(input));
+        }
+        return least(values);
+    }
+
+    private static Object getMinOrMax(Collection<Object> values, BiFunction<Object, Object, Boolean> comparison) {
```
[the remark this answers is not in the document]
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/Conditionals.java]: But the arguments can also be Strings.

**Comment by matriv** [comment]
@costin Thanks for the review. Pushed commits to address comments.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/Foldables.java [hunk 3]
```diff
@@ -46,11 +49,18 @@ public static double doubleValueOf(Expression e) {
...
-        for (Expression e : list) {
-             l.add(valueOf(e, to));
+        return foldAndAdd(list, to, new ArrayList<>(list.size()));
+    }
+
+    public static <T> Set<T> valuesOfNoDuplicates(List<Expression> list, DataType to) {
+        return foldAndAdd(list, to, new LinkedHashSet<>(list.size()));
+    }
+
+    private static <T, C extends Collection<T>> C foldAndAdd(Collection<Expression> expressions, DataType to, C values) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/Foldables.java]: Please rename it to `foldTo` - I was thinking the values are folded and added together based on the current name.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/ArbitraryConditionalFunction.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.StringJoiner;
+
+import static org.elasticsearch.xpack.sql.expression.gen.script.ParamsBuilder.paramsBuilder;
+
+/**
+ * Base class for conditional predicates with arbitrary number of arguments
+ */
+public abstract class ArbitraryConditionalFunction extends ConditionalFunction {
+
+    private ConditionalOperation operation;
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/ArbitraryConditionalFunction.java]: `final`

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/Conditionals.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        for (Object object : values) {
+            if (object != null) {
+                return object;
+            }
+        }
+
+        return null;
+    }
+
+    static Object coalesceOnInput(List<Processor> processors, Object input) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/Conditionals.java]: The `On` is superfluous. - `coalesceInput`, `greatestInput`, etc...

**Review by costin — APPROVED** [review]
Left some minor polishing comments.
LGTM!

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/Conditionals.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            return null;
+        }
+
+        Object result = null;
+        int i = 0;
+        for (Object value : values) {
+            if ((i == 0) || (result == null) || (comparison.apply(value, result) == Boolean.TRUE)) {
+                result = value;
+            }
+            i++;
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/Conditionals.java]: I think here the `i` logic can be replaced by a `boolean`? Seems to be a `true/false` scenario.

**Review by astefan — APPROVED** [review]
LGTM. Left one question only.

**Comment by matriv** [comment]
Backported to `6.x` with 44d6c74214ab784ad324e7b5ecbb56d18940be8c

**Pull request merged**